Updation of import-restrictions file

yashsingh74 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
The import-boss script file requires the `.import-restrictions` file. This will help the maintainers keep the dependencies clean. It is described as a YAML or JSON format. This file is not included in the CI runs. When it is added, the list of directories and subdirectories is also added.


As there is a list of directories and subdirectories that do not exist in the k/k repo. This PR helps to clean up the path to those directories and adds a list of directories that exist in the k/k path.

Ref: [import-boss](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/code-generator/cmd/import-boss)
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #115234

Commit:
Cleans up the path to those directories and adds a list of directories that exist in the k/k path.

Signed-off-by: Yash Singh <syash@vmware.com>

yashsingh74 commented:
@pohly Please review the PR.

pohly reviewed (changes requested):
Can you add comments explaining why you are making these changes and why these (and just these entries) are correct? Start commenting the file and then add further information in the commit message.

Assume that I don't have a clue how import-boss works (because I don't...).

Also note that there is a TODO in the file. Can that be resolved?

yashsingh74 commented on test/e2e/framework/.import-restrictions:
```diff
@@ -1,116 +1,454 @@
 rules:
   - selectorRegexp: k8s[.]io/kubernetes/pkg/
```
import boss checks the packages in place using this format. Since just the packages used by the framework are being imported, not all of the k/k packages.
Also, we are using the defined path to import the pkg and test path instead of the 'k8s[.]io' parent path.

pohly replied:
By looking just at this file and PR, it's still not clear to me what is being matched and what the effect is. Can you perhaps add comments to the file?

I probably also have to read up on how import boss works, but even then having those comments in the file will help the next person reading it.

yashsingh74 replied:
I have updated which packages are used in the test and fixes the external Packages which are used in the test.
Also, the packages only includes the allowed prefix.

pohly replied:
I checked https://pkg.go.dev/k8s.io/gengo/examples/import-boss, but I am still trying to wrap my head around it.

So here we make a statement about importing some package under k8s[.]io/kubernetes/pkg: only the packages listed explicitly are allowed?

This is a surprisingly long list - much too long IMHO. I suspect this is a result of using this `.import-restrictions` file for the core framework *and* for the domain-specific sub-packages.

I believe what we want is a fairly tight list here and then individual `.import-restrictions` files in the sub-package were we allow additional packages on a per-sub-package basis.

When you extended this list, how did you decide which packages to list? Are all of them really needed somewhere or did you list everything that currently exists?

pohly commented on test/e2e/framework/.import-restrictions:
```diff
@@ -1,269 +1,220 @@
+  # Packages are used in the test/
   - selectorRegexp: k8s[.]io/kubernetes/test/
     allowedPrefixes:
+      - k8s.io/kubernetes/test/conformance/testdata
```
I don't think we need to (and shouldn't) list every single package here.

What we want to say is "anything under test/e2e/framework may import anything under test". Or at least I can't think of something that shouldn't be allowed.

If that is the goal, then can't we just use this:
```
  # Everything under k8s.io/kubernetes/test/ may be used.
  - selectorRegexp: k8s[.]io/kubernetes/test/
    allowedPrefixes:
    - ""
```

Or even leave the rule out entirely?

yashsingh74 replied:
According to the first comment, I have listed down all the packages which are being used in the test package. Yes, the list is too long so it is better to only keep prefix which are being used.

Second, in the */test package we can use the suggested changes as most of the import are used.

I will make the changes.

pohly commented on test/e2e/framework/.import-restrictions:
```diff
@@ -1,269 +1,220 @@
   - selectorRegexp: k8s[.]io/kubernetes/third_party/
     allowedPrefixes:
-    - k8s.io/kubernetes/third_party/forked/golang/expansion
+    - k8s.io/kubernetes/third_party/forked/golang/net
```
According to my comment above, this would belong into `test/e2e/network/.import-restrictions`.

yashsingh74 replied:
Done.

Commit:
Updated the file with the suggested changes

Commit:
Updated the file with the suggested changes

pohly commented on test/e2e/framework/.import-restrictions:
```diff
@@ -1,269 +1,23 @@
-  - selectorRegexp: k8s[.]io/kubernetes/third_party/
-    allowedPrefixes:
-    - k8s.io/kubernetes/third_party/forked/golang/expansion
+      - ""
```
This rule can be removed, it doesn't do anything.

pohly commented on test/e2e/framework/.import-restrictions:
```diff
@@ -1,269 +1,23 @@
 rules:
+  # prevent any k8s.io/kubernetes/pkg imports outside of this package
+  # Packages are used in the test
```
```suggestion
  # The core E2E framework is meant to be normal Kubernetes client,
  # which means that it shouldn't depend on internal
  # code. But we are not there yet, so some exceptions
  # have to be allowed. Over time the list of allowed
  # packages should get shorter, not longer.
```

pohly reviewed (changes requested):
I tried out `hack/verify-import-boss.sh` with your branch and it is failing for me:
```
E0207 17:25:02.971904   76898 main.go:41] Error: Failed executing generator: some packages had errors:
errors in package "k8s.io/kubernetes/test/e2e/framework/providers/aws":
the following imports did not match any allowed prefix:
  k8s.io/kubernetes/pkg/api/legacyscheme
  k8s.io/kubernetes/pkg/apis/apps
  k8s.io/kubernetes/pkg/apis/autoscaling
...
```

Same for several sub-packages.

I think we need to override the restrictions for k8s.io/kubernetes/pkg for sub-packages.

If I add the following file to several directories, the check passes:
```
rules:
  # Allow full usage of the entire k8s.io/kubernetes/pkg.
  # This is probably too broad and could be restricted further,
  # but that isn't important. What matters are the restrictions
  # of the core framework: those need to be overridden here
  # due to how the E2E packages are structured.
  - selectorRegexp: k8s[.]io/kubernetes/pkg/
    allowedPrefixes:
    - ""
```

```
	test/e2e/framework/daemonset/.import-restrictions
	test/e2e/framework/kubelet/.import-restrictions
	test/e2e/framework/perf/.import-restrictions
	test/e2e/framework/providers/aws/.import-restrictions
	test/e2e/framework/providers/gce/.import-restrictions
	test/e2e/framework/providers/kubemark/.import-restrictions
	test/e2e/framework/pv/.import-restrictions
```

Commit:
Added the suggested changes

yashsingh74 commented:
@pohly I have updated the file and ran the `hack/verify-import-boss.sh` script. Now it doesn't show the same error.
The reason I included the allowedPrefix imports is because it is used in the test/e2e/framework. Also, removed the `k8s[.]io/kubernetes/test/` selector.
Please confirm and let me know if it requires any more changes.

pohly commented:
/ok-to-test

Commit:
Added the suggested changes

pohly commented:
This is still not what I was aimining at:
- test/e2e/framework/*.go should have very minimal dependencies. We need to enforce that via import-boss.
- What each test/e2e/framework/* sub-package uses is less relevant, although ideally it also should be as minimal as possible in each case.

Now that I understand how import-boss works, it's easier for me to spell it out with code than with words :cry: 

Sorry, but I think I'll solve https://github.com/kubernetes/kubernetes/issues/115234 myself. Can you perhaps review https://github.com/kubernetes/kubernetes/pull/115710?

yashsingh74 commented:
Thank you @pohly I will look up for your open PR.

pohly commented:
/close 

https://github.com/kubernetes/kubernetes/pull/115710 was merged.
